### ISSUE TITLE
`--fp16` causing loss to go to Inf or NaN

### ISSUE THREAD

**Issue by afiaka87** [lucidrains/DALLE-pytorch#169]
This is an error I received a few days ago. Haven't been able to try it on the newest release so do let me know if you found out what was going on already.

Anyway yeah it seems a division is perhaps causing an underflow or something of that nature with higher batch sizes - not sure exactly. I have a W&B report that I will link shortly to an example run which produced this issue.

**Comment by lucidrains** [comment]
@afiaka87 oh oops! i realized deepspeed handles gradient clipping, made the change https://github.com/lucidrains/DALLE-pytorch/blob/main/train_dalle.py#L278 (it may or may not make a difference)

**Comment by afiaka87** [comment]
Hopefully that fixed it! Won't be able to test with those batch sizes myself unfortunately. Not today anyway. I'll try to remember to check this again though.

**Comment by janEbert** [comment]
This is just a general problem with fp16. It's much more sensitive to the parameters. The gradient clipping change with DeepSpeed is great, though, lets us clean the code up a bit more!

**Comment by janEbert** [comment]
This is also the reason why I haven't PR'd [fp16 for train_vae.py](https://github.com/janEbert/DALLE-pytorch/commit/39f49fe31ae8bb55cab423f99c153381db4252e1) – the default parameters there lead to the same issue.

**Comment by robvanvolt** [comment]
I am currently training with deepspeed and --fp16 and taming transformer, and interestingly, the loss gets infinite if the batch_size is set too high (below the VRAM limit of course)... interesting!

**Comment by afiaka87** [comment]
Still having this issue unfortunately. Going to tinker with some deepspeed config settings to see what works.

**Comment by janEbert** [comment]
Since the discussion revolves so heavily around DeepSpeed, I feel it's important to remind you that the same behaviour can be seen without DeepSpeed as well.
Changing DeepSpeed's parameters may be able to fix this since you can't do the same in PyTorch so easily; but in the end, it's the training parameters that cause these issues to appear in the first place.

**Comment by afiaka87** [comment]
For people training with fp16:
- OpenAI tried and they had a ton of trouble getting it to work
- Consider using horovod with automatic mixed precision instead. 
- If you're on a single GPU - use deepspeed's amp config (uses Nvidia apex under the hood) 

Thanks @janEbert - i hadn't done the due diligence needed to understand this was such a task. I do think adding a "skip up to 5% of iterations if they give Inf NaN values" feature would be a simple fix which they suggest in the paper.

 I don't know that I could recommend using fp16 even with that fix in place though now that I know enough. My biggest concern is that a lot of people training are sort of "VRAM cutters" - basically increasing batch size and toggling train_dalle.py `arg` until they can increase batch size again. Rinse, wash, repeat. They turn on fp16 and get something like 25%-50% memory savings basically for free and assume it's going to help when they can't _actually_ increase the batch size anymore and may even have to decrease their batch size to stabilize training. At that point there's really no reason not to just use fp32 - it's better and will happily use up all your VRAM without having over/underflow issues. 

It's definitely impressive work they did to find the correct hyperparameters for working with fp16 but I think implementing all of their changes would need to happen at the network level - which I don't think I'm yet qualified to figure out. We have a reversible transformer and the 1024 token VQGAN though - they didn't. 

@lucidrains @janEbert I have neither the motivation nor the skillset to implement these changes. Do either of you? Otherwise I'm just gonna close this I think. Maybe we could print a message to this issue's url for people training with fp16? And do the loss skip thing?

**Comment by afiaka87** [comment]
> This is also the reason why I haven't PR'd [fp16 for train_vae.py](https://github.com/janEbert/DALLE-pytorch/commit/39f49fe31ae8bb55cab423f99c153381db4252e1) – the default parameters there lead to the same issue.

Again - sorry I didn't follow up on this well enough. I guess this wound up in there despite your protest? I agree now - it's worth considering removal from the default train_dalle.py. People who really need it can lookup the deepspeed config for it on the wiki.
